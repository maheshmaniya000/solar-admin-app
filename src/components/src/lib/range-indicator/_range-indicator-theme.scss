@use '~@angular/material' as mat;

@mixin theme($theme-or-color-config) {
  $color-config: mat.get-color-config($theme-or-color-config);
  $background: map-get($color-config, background);
  $foreground: map-get($color-config, foreground);

  sg-range-indicator {
    background-color: mat.get-color-from-palette($background, status-bar);

    .target {
      background-color: mat.get-color-from-palette($foreground, divider);
    }
  }
}
