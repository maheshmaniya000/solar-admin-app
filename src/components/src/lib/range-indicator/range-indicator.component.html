<div
  [style.width.%]="barWidth"
  [ngClass]="barColorClass"
  aria-readonly="true"
  role="slider"
  [attr.aria-valuemin]="min"
  [attr.aria-valuemax]="max"
  [attr.aria-valuenow]="value"
></div>
<div class="target" [style.left.%]="targetPosition"></div>
