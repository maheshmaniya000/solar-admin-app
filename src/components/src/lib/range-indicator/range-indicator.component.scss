@use 'variables';

:host {
  position: relative;
  display: inline-block;
  width: 100%;
  height: 6px;
}

div {
  position: absolute;
  height: 100%;
  transition: variables.$sg-slider-transition;
}

.less {
  background-color: map-get(variables.$sg-status-colors, warning);
}

.equal {
  background-color: map-get(variables.$sg-status-colors, success);
}

.greater {
  background-color: map-get(variables.$sg-status-colors, error);
}

.target {
  width: 1px;
}
