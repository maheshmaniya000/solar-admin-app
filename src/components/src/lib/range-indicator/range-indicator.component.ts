import {
  ChangeDetectionStrategy,
  Component,
  Input,
  OnChanges
} from '@angular/core';
import { clamp } from 'lodash-es';

import { calculatePercentageWithinRange } from '../utils';

@Component({
  selector: 'sg-range-indicator',
  templateUrl: './range-indicator.component.html',
  styleUrls: ['./range-indicator.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class RangeIndicatorComponent implements OnChanges {
  @Input() min: number;
  @Input() max: number;
  @Input() target: number;
  @Input() value: number;

  barWidth: number;
  targetPosition: number;
  barColorClass: string;

  ngOnChanges(): void {
    if (this.max <= this.min) {
      throw new Error(
        `Invalid inputs for RangeIndicatorComponent: min must be less than max (min: ${this.min}, max: ${this.max})`
      );
    }
    this.value = clamp(this.value, this.min, this.max);
    this.barWidth = calculatePercentageWithinRange(
      this.value,
      this.min,
      this.max
    );
    this.targetPosition = calculatePercentageWithinRange(
      this.target,
      this.min,
      this.max
    );
    this.barColorClass = this.getBarColorClass();
  }

  private getBarColorClass(): string {
    return this.value < this.target
      ? 'less'
      : this.value === this.target
      ? 'equal'
      : 'greater';
  }
}
