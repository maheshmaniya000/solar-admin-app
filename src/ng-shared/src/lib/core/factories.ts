
export function windowFactory(): any { return window; }

export function appBaseHrefFactory(app: string): string {
  return `/${app}/`;
}
