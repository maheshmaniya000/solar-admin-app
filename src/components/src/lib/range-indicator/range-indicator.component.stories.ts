import { moduleMetadata, Story } from '@storybook/angular';

import { RangeIndicatorComponent } from './range-indicator.component';
import { RangeIndicatorModule } from './range-indicator.module';

const styles = [
  `
    .small {
      width: 100px;
      display: block;
      margin: 1em;
    }
    .big {
      width: 500px;
      height: 20px;
      display: block;
      margin: 1em;
    }
  `
];

export default {
  title: 'Solargis/Components/Range Indicator',
  component: RangeIndicatorComponent,
  decorators: [
    moduleMetadata({
      imports: [RangeIndicatorModule]
    })
  ]
};

const createTemplate = (styleClass: string): string => `
  <sg-range-indicator class="${styleClass}" [min]="0" [max]="100" [target]="80" [value]="50"></sg-range-indicator>
  <sg-range-indicator class="${styleClass}" [min]="0" [max]="100" [target]="80" [value]="90"></sg-range-indicator>
  <sg-range-indicator class="${styleClass}" [min]="0" [max]="100" [target]="80" [value]="80"></sg-range-indicator>
`;

export const small: Story = () => ({
  template: createTemplate('small'),
  styles
});

export const large: Story = () => ({
  template: createTemplate('big'),
  styles
});

export const interactive: Story = props => ({
  props,
  styles
});

interactive.argTypes = {
  min: {
    name: 'Minimum',
    control: { type: 'number' },
    defaultValue: 0
  },
  max: {
    name: 'Maximum',
    control: { type: 'number' },
    defaultValue: 100
  },
  value: {
    name: 'Value',
    control: { type: 'number' },
    defaultValue: 50
  },
  target: {
    name: 'Target',
    control: { type: 'number' },
    defaultValue: 80
  }
};
