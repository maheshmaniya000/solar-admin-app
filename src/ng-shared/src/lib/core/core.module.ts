import { APP_BASE_HREF } from '@angular/common';
import { HttpClientJsonpModule, HttpClientModule, HTTP_INTERCEPTORS }  from '@angular/common/http';
import { APP_BOOTSTRAP_LISTENER, ErrorHandler, Inject, NgModule, Optional, SkipSelf } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { MatBadgeModule } from '@angular/material/badge';
import { MatButtonModule } from '@angular/material/button';
import { MatCardModule } from '@angular/material/card';
import { MatDialogModule } from '@angular/material/dialog';
import { MatDividerModule } from '@angular/material/divider';
import { MatIconModule, MatIconRegistry } from '@angular/material/icon';
import { MatInputModule } from '@angular/material/input';
import { MatListModule } from '@angular/material/list';
import { MatMenuModule } from '@angular/material/menu';
import { MatPaginatorIntl, MatPaginatorModule } from '@angular/material/paginator';
import { MatProgressSpinnerModule } from '@angular/material/progress-spinner';
import { MatTabsModule } from '@angular/material/tabs';
import { MatToolbarModule } from '@angular/material/toolbar';
import { DomSanitizer } from '@angular/platform-browser';
import { Router, RouteReuseStrategy, UrlSerializer } from '@angular/router';
import { TranslocoService } from '@ngneat/transloco';
import { EffectsModule } from '@ngrx/effects';
import { RouterStateSerializer } from '@ngrx/router-store';
import { Store } from '@ngrx/store';

import { analyticsFactory, AnalyticsService } from '@solargis/ng-analytics';
import { GEOCODER_CONFIG_TOKEN, GeocoderConfig, GeosearchModule } from '@solargis/ng-geosearch';
import { UnitToggleService, UnitValuePipe } from '@solargis/ng-unit-value';

import { TranslocoRootModule } from 'ng-shared/core/transloco-root.module';
import { UserSharedModule } from 'ng-shared/user-shared/user-shared.module';

import { Config } from '../config';
import { SharedModule } from '../shared/shared.module';
import { UserModule } from '../user/user.module';
import { CustomReuseStrategy } from '../utils/router/custom.reuse-stragegy';
import { CustomRouterStateSerializer } from '../utils/router/custom.router-state.serializer';
import { CustomUrlSerializer } from '../utils/router/custom.url-serializer';
import { AfterBootstrap } from './actions/bootstrap.action';
import { AppMenuComponent } from './components/app-menu/app-menu.component';
import { AcceptCookiesComponent } from './components/cookies/accept-cookies.component';
import { HeaderComponent } from './components/header/header.component';
import { MobileMenuComponent } from './components/mobile-menu/mobile-menu.component';
import { AppComponent } from './containers/app.component';
import { ContentLockedDialogComponent } from './dialogs/content-locked-dialog/content-locked-dialog.component';
import { AmplitudeEffects } from './effects/amplitude.effects';
import { ConfigEffects } from './effects/config.effects';
import { DialogEffects } from './effects/dialog.effects';
import { LayoutEffects } from './effects/layout.effects';
import { RouteEffects } from './effects/route.effects';
import { RouteDataEffects } from './effects/router-data.effects';
import { SettingsEffects } from './effects/settings.effects';
import { TranslateEffects } from './effects/translate.effects';
import { UrlParamsEffects } from './effects/url-params.effects';
import { appBaseHrefFactory, windowFactory } from './factories';
import { CompanyTokenInterceptor } from './interceptors/company-token.interceptor';
import { IdTokenInterceptor } from './interceptors/id-token.interceptor';
import { UrlEncodeInterceptor } from './interceptors/urlencode.interceptor';
import { getReducers } from './reducers';
import { DateTimeFormatNgrxService } from './services/date-time-format-ngrx.service';
import { DateTimeFormatService } from './services/date-time-format.service';
import { GeolocatorService } from './services/geolocator.service';
import { GlobalErrorHandler } from './services/global.error-handler';
import { SgPaginatorIntl } from './services/sg-paginator-intl';
import { UnitToggleNgrxService } from './services/unit-toggle-ngrx.service';
import { APP_TOKEN, REDUCER_TOKEN } from './tokens';

function throwIfAlreadyLoaded(parentModule: any, moduleName: string): void {
  if (parentModule) {
    throw new Error(`${moduleName} has already been loaded. Import Core modules in the AppModule only.`);
  }
}

export function appBootstrapListenerFactory(store: Store<any>): () => void {
  return () => {
    store.dispatch(new AfterBootstrap());
  };
}

export function geocoderConfigFactory(config: Config, transloco: TranslocoService): GeocoderConfig {
  return {
    ...config.geocoder,
    language: 'en',
    getLanguageOnEachRequest: () => transloco.getActiveLang()
  };
}

@NgModule({
  imports: [
    HttpClientModule,
    HttpClientJsonpModule,
    EffectsModule.forFeature([
      UrlParamsEffects,
      RouteEffects,
      RouteDataEffects,
      SettingsEffects,
      TranslateEffects,
      LayoutEffects,
      AmplitudeEffects,
      DialogEffects,
      ConfigEffects
    ]),
    FormsModule,
    GeosearchModule.forRoot({
      provide: GEOCODER_CONFIG_TOKEN,
      useFactory: geocoderConfigFactory,
      deps: [Config, TranslocoService]
    }),
    MatBadgeModule,
    MatButtonModule,
    MatCardModule,
    MatDialogModule,
    MatDividerModule,
    MatIconModule,
    MatInputModule,
    MatListModule,
    MatMenuModule,
    MatPaginatorModule,
    MatProgressSpinnerModule,
    MatTabsModule,
    MatToolbarModule,
    SharedModule,
    TranslocoRootModule,
    UserModule,
    UserSharedModule
  ],
  declarations: [
    AcceptCookiesComponent,
    AppComponent,
    AppMenuComponent,
    ContentLockedDialogComponent,
    HeaderComponent,
    MobileMenuComponent,
  ],
  exports: [AppComponent],
  entryComponents: [
    ContentLockedDialogComponent
  ],
  providers: [
    // solargis services
    // provided in root: AmplitudeAnalyticsService,
    GeolocatorService,
    // provided in root: StorageProviderService,
    UnitValuePipe,
    { provide: AnalyticsService, useFactory: analyticsFactory, deps: [Config, Router, 'Window'] },
    { provide: ErrorHandler, useClass: GlobalErrorHandler },
    { provide: MatPaginatorIntl, useClass: SgPaginatorIntl, deps: [TranslocoService] },
    { provide: UnitToggleService, useClass: UnitToggleNgrxService, deps: [Store] },
    { provide: DateTimeFormatService, useClass: DateTimeFormatNgrxService, deps: [Store] },
    // angular router
    { provide: RouteReuseStrategy, useClass: CustomReuseStrategy },
    { provide: RouterStateSerializer, useClass: CustomRouterStateSerializer },
    { provide: UrlSerializer, useClass: CustomUrlSerializer },
    // Ngrx store
    { provide: REDUCER_TOKEN, useFactory: getReducers },
    // App config and bootstrap
    { provide: APP_BASE_HREF, useFactory: appBaseHrefFactory, deps: [APP_TOKEN] },
    { provide: APP_BOOTSTRAP_LISTENER, useFactory: appBootstrapListenerFactory, deps: [Store], multi: true },
    // Interceptors
    { provide: HTTP_INTERCEPTORS, useClass: CompanyTokenInterceptor, multi: true },
    { provide: HTTP_INTERCEPTORS, useClass: IdTokenInterceptor, multi: true },
    { provide: HTTP_INTERCEPTORS, useClass: UrlEncodeInterceptor, multi: true },
    // other
    { provide: 'Window', useFactory: windowFactory },
  ]
})
export class CoreModule {

  constructor(
    @Optional() @SkipSelf() parentModule: CoreModule,
    @Inject(APP_BASE_HREF) baseHref: string,
    iconRegistry: MatIconRegistry,
    sanitizer: DomSanitizer
  ) {
    throwIfAlreadyLoaded(parentModule, 'CoreModule');
    iconRegistry.addSvgIconSetInNamespace(
      'sg',
      sanitizer.bypassSecurityTrustResourceUrl(baseHref + 'assets/img/solargis-icons.svg')
    );
  }

}
