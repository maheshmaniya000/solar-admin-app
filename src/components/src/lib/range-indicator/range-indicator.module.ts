import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';

import { SgDefaultsModule } from '../defaults/defaults.module';
import { RangeIndicatorComponent } from './range-indicator.component';

@NgModule({
  declarations: [RangeIndicatorComponent],
  imports: [CommonModule, SgDefaultsModule],
  exports: [RangeIndicatorComponent]
})
export class RangeIndicatorModule {}
