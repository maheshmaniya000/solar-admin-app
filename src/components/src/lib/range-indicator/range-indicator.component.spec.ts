import { ComponentFixture, TestBed, waitForAsync } from '@angular/core/testing';

import { RangeIndicatorComponent } from './range-indicator.component';
import { RangeIndicatorModule } from './range-indicator.module';

interface RangeIndicatorConfig {
  min: number;
  max: number;
  target: number;
  value: number;
}

describe('RangeIndicatorComponent', () => {
  const setInputs = (
    { min, max, target, value }: RangeIndicatorConfig,
    component: RangeIndicatorComponent
  ): void => {
    component.min = min;
    component.max = max;
    component.target = target;
    component.value = value;
    component.ngOnChanges();
  };

  describe('unit', () => {
    let component: RangeIndicatorComponent;

    const setInputsForUnitTest = (config: RangeIndicatorConfig): void => {
      setInputs(config, component);
    };

    beforeEach(() => {
      component = new RangeIndicatorComponent();
    });

    describe('ngOnChanges', () => {
      it('should throw when `min` is greater than `max`', () => {
        expect(() =>
          setInputsForUnitTest({ min: 150, max: 100, target: 80, value: 50 })
        ).toThrow();
      });

      it('should throw when `min` is equal to `max`', () => {
        expect(() =>
          setInputsForUnitTest({ min: 0, max: 0, target: 0, value: 0 })
        ).toThrow();
      });
    });

    describe('barWidth', () => {
      it('should be correctly recalculated when inputs change', () => {
        setInputsForUnitTest({ min: 0, max: 100, target: 80, value: 50 });
        expect(component.barWidth).toBe(50);

        setInputsForUnitTest({ min: -50, max: 50, target: 20, value: 10 });
        expect(component.barWidth).toBe(60);
      });

      it('should be 0 when `value` is less than `min`', () => {
        setInputsForUnitTest({ min: 0, max: 100, target: 80, value: -1 });
        expect(component.barWidth).toBe(0);
      });

      it('should be 100 when `value` is greater than `max`', () => {
        setInputsForUnitTest({ min: 0, max: 100, target: 80, value: 101 });
        expect(component.barWidth).toBe(100);
      });
    });

    describe('targetPosition', () => {
      it('should be correctly recalculated when inputs change', () => {
        setInputsForUnitTest({ min: 0, max: 100, target: 80, value: 50 });
        expect(component.targetPosition).toBe(80);

        setInputsForUnitTest({ min: -50, max: 50, target: -40, value: 50 });
        expect(component.targetPosition).toBe(10);
      });

      it('should be 0 when `target` is less than `min`', () => {
        setInputsForUnitTest({ min: 0, max: 100, target: -1, value: 50 });
        expect(component.targetPosition).toBe(0);
      });

      it('should be 100 when `target` is greater than `min`', () => {
        setInputsForUnitTest({ min: 0, max: 100, target: 101, value: 50 });
        expect(component.targetPosition).toBe(100);
      });
    });

    describe('barColorClass', () => {
      it('should be `less` when `value` is less than `target`', () => {
        setInputsForUnitTest({ min: 0, max: 100, target: 51, value: 50 });
        expect(component.barColorClass).toBe('less');
      });

      it('should be `equal` when `value` is equal `target`', () => {
        setInputsForUnitTest({ min: 0, max: 100, target: 50, value: 50 });
        expect(component.barColorClass).toBe('equal');
      });

      it('should be `greater` when `value` is greater than `target`', () => {
        setInputsForUnitTest({ min: 0, max: 100, target: 49, value: 50 });
        expect(component.barColorClass).toBe('greater');
      });
    });
  });

  describe('component', () => {
    let fixture: ComponentFixture<RangeIndicatorComponent>;
    let component: RangeIndicatorComponent;

    const setInputsForComponentTest = (config: RangeIndicatorConfig): void => {
      setInputs(config, component);
      fixture.detectChanges();
    };

    beforeEach(
      waitForAsync(() => {
        TestBed.configureTestingModule({
          imports: [RangeIndicatorModule]
        }).compileComponents();

        fixture = TestBed.createComponent(RangeIndicatorComponent);
        component = fixture.componentInstance;
      })
    );

    it('should correctly render when `value` is less than `target`', () => {
      setInputsForComponentTest({ min: 0, max: 100, value: 50, target: 80 });
      expect(fixture).toMatchSnapshot();
    });

    it('should correctly render when `value` is equal to `target`', () => {
      setInputsForComponentTest({ min: 0, max: 100, value: 50, target: 50 });
      expect(fixture).toMatchSnapshot();
    });

    it('should correctly render when `value` is greater than `target`', () => {
      setInputsForComponentTest({ min: 0, max: 100, value: 50, target: 20 });
      expect(fixture).toMatchSnapshot();
    });
  });
});
