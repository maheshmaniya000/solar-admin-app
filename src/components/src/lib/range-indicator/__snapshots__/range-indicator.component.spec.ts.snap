// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RangeIndicatorComponent component should correctly render when \`value\` is equal to \`target\` 1`] = `
<sg-range-indicator
  __ngContext__={[Function LRootView]}
  barColorClass={[Function String]}
  barWidth={[Function Number]}
  max={[Function Number]}
  min="0"
  target={[Function Number]}
  targetPosition={[Function Number]}
  value={[Function Number]}
>
  <div
    aria-readonly="true"
    aria-valuemax="100"
    aria-valuemin="0"
    aria-valuenow="50"
    class="equal"
    role="slider"
    style="width: 50%;"
  /><div
    class="target"
    style="left: 50%;"
  />
</sg-range-indicator>
`;

exports[`RangeIndicatorComponent component should correctly render when \`value\` is greater than \`target\` 1`] = `
<sg-range-indicator
  __ngContext__={[Function LRootView]}
  barColorClass={[Function String]}
  barWidth={[Function Number]}
  max={[Function Number]}
  min="0"
  target={[Function Number]}
  targetPosition={[Function Number]}
  value={[Function Number]}
>
  <div
    aria-readonly="true"
    aria-valuemax="100"
    aria-valuemin="0"
    aria-valuenow="50"
    class="greater"
    role="slider"
    style="width: 50%;"
  /><div
    class="target"
    style="left: 20%;"
  />
</sg-range-indicator>
`;

exports[`RangeIndicatorComponent component should correctly render when \`value\` is less than \`target\` 1`] = `
<sg-range-indicator
  __ngContext__={[Function LRootView]}
  barColorClass={[Function String]}
  barWidth={[Function Number]}
  max={[Function Number]}
  min="0"
  target={[Function Number]}
  targetPosition={[Function Number]}
  value={[Function Number]}
>
  <div
    aria-readonly="true"
    aria-valuemax="100"
    aria-valuemin="0"
    aria-valuenow="50"
    class="less"
    role="slider"
    style="width: 50%;"
  /><div
    class="target"
    style="left: 80%;"
  />
</sg-range-indicator>
`;
